### PR TITLE
[fix] ReduceOps throw error if dim is repeated

### LINKED ISSUE #44273: mean with repeated dim gives inconsistent results for cpu and cuda
## To Reproduce

```
import torch
a = torch.arange(1,17).float().reshape(4,4)
print(a)
cpu_res = torch.mean(a, (0,0))
cuda_res = torch.mean(a.cuda(), (0,0))
print(cpu_res)
print(cuda_res)
```
Result is 
```
tensor([[ 1.,  2.,  3.,  4.],
        [ 5.,  6.,  7.,  8.],
        [ 9., 10., 11., 12.],
        [13., 14., 15., 16.]])
tensor([1.7500, 2.0000, 2.2500, 2.5000])
tensor([ 7.,  8.,  9., 10.], device='cuda:0')
```

When transfer a same dim like 0 twice, CUDA and CPU return different result. It seems like CPU calculates mean twice and CUDA not.

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #44273 

TODO

* [x] Add test

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Should this be made in some reusable helper? So it's more discoverable if a similar check is decided for other ops

**Comment by kshitij12345** [comment]
> Should this be made into some reusable helper? So it's more discoverable if a similar check is decided for other ops

Makes sense.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -358,6 +358,9 @@ Tensor& prod_out(Tensor& result, const Tensor& self, Dimname dim,
 
 Tensor &mean_out_cpu_gpu(Tensor &result, const Tensor &self, IntArrayRef dim,
                  bool keepdim, c10::optional<ScalarType> opt_dtype) {
+  auto dim_set = std::set<IntArrayRef::value_type>(dim.cbegin(), dim.cend());
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOps.cpp:361]: This is probably not actually the most efficient way to do the duplicate check, since it involves doing a fairly unnecessary dynamic allocation for the set. Probably quickest when number of dims is small (which it should be usually) is just the quadratic nested loops version.

File: test/test_torch.py [hunk 2]
```diff
@@ -18859,6 +18859,10 @@ def test_dstack(self, device, dtype):
             expected = np.dstack(np_input)
             self.assertEqual(actual, expected)
 
+    def test_mean_repeated_dim(self, device):
+        x = torch.randn(3, 3, 3, 3, device=device)
+        with self.assertRaisesRegex(RuntimeError, r'mean: repeated dimension in `dim` \(\[0, 0\]\)'):
+            torch.mean(x, dim=(0, 0))
```
Comment (ezyang) [inline test/test_torch.py:18865]: @mruberry So, as was mentioned in the original issue, there are a bunch of operators which take in a list of dimensions. It seems like it would be useful to easily run a version of this test for all of the operators that do this :>

**Review by ezyang — APPROVED** [review]

File: test/test_torch.py (hunk 2 above)
Comment (mruberry) [reply test/test_torch.py:18865]: Good idea. I've made a note.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/ReduceOps.cpp:361]: Makes sense.
What I had in mind was something like
```c++
if (dim.size() < 10) { // 10 is just some heurestic value
   // for loop version
} else {
  // set version.
}
```
Let me know if it sounds good and if you approve it what should the heuristic value be?

Thank You!

File: test/test_torch.py (hunk 2 above)
Comment (kshitij12345) [reply test/test_torch.py:18865]: I had quickly searched the file below
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml
for `int[1]` based on the signature of https://github.com/pytorch/pytorch/blob/cce5982c4ceb77a0797e8bd4c717ebfec0681eab/aten/src/ATen/native/native_functions.yaml#L2093

**Comment by ezyang** [comment]
Ooh, this new version is much better, thanks!

**Comment by ezyang** [comment]
cc @ailzhang on the xla bit

**Comment by kshitij12345** [comment]
Gentle Ping :)
